irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -15,103 +15,118 @@
-                  proposal_set_.erase(hash(tx).to_string());
-                  candidate_set_.insert(hash(tx).to_string());
-                  TransactionResponse response;
-                  response.tx_hash = hash(tx).to_string();
-                  response.current_status =
-                      model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS;
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto block) {
+              static const auto in_proposal = [this](const auto &tx) {
+                return this->proposal_set_.count(hash(tx).to_string());
+              };
+              boost::for_each(
+                  block.transactions | boost::adaptors::filtered(in_proposal),
+                  notify_success);
             },
             // on complete
             [this]() {
-              for (auto tx_hash : proposal_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status =
-                    TransactionResponse::STATEFUL_VALIDATION_FAILED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              proposal_set_.clear();
-
-              for (auto tx_hash : candidate_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status = TransactionResponse::COMMITTED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              candidate_set_.clear();
+              boost::for_each(
+                  boost::join(this->proposal_set_, this->candidate_set_),
+                  notify_fail);
+              this->proposal_set_.clear();
+              this->candidate_set_.clear();
             });
       });
+
+      mst_proc_->onPreparedTransactions().subscribe(
+          [](auto tx) { return notify_success(*tx); });
+      mst_proc_->onExpiredTransactions().subscribe(
+          [](auto tx) { return notify_fail(hash(*tx).to_string()); });
     }
 
     void TransactionProcessorImpl::transactionHandle(
-        std::shared_ptr<model::Transaction> transaction) {
+        ConstRefTransaction transaction) {
       log_->info("handle transaction");
-      model::TransactionResponse response;
-      response.tx_hash = hash(*transaction).to_string();
-      response.current_status =
-          model::TransactionResponse::Status::STATELESS_VALIDATION_FAILED;
+      TransactionResponse response{hash(*transaction).to_string(),
+                                   Status::STATELESS_VALIDATION_FAILED};
 
+      // TODO: nice place for code linearizing
       if (validator_->validate(*transaction)) {
```

Comment:
Also, feel that providing information about MST transactions it is not a responsibility of validator. Need to move this business logic to tx processor.